### PR TITLE
Implement GetWorkerTaskReachability API

### LINKED ISSUE #3303: Worker Versioning: Add visibility tags (and possibly API) to enable worker retirement discovery
We need a way for users to be able to understand when a set of workers can be retired because no more workflows require the version those workers support.

We'll also likely eventually use this to provide automatic notifications of some sort.

We could accomplish this more than one way, but, like we've done with the proto messages it probably makes sense to avoid duplicating the data in the existing `BinaryChecksums` search attribute while adding a new `BuildIDs` search attribute.

That also said, `BuildIDs` would be a much preferable name. After a little googling, seems ES supports aliases, so that seems a good direction.

It's also not clear that the set of all build ids which the workflow has used is sufficient for a user to simply issue a visibility query and learn everything they need to know. At minimum, you'd also need to know which version the _last_ wft was processed with, but even that is likely insufficient. EX: Last WFT was processed with `2.1` but the graph contains `2.2`. You could query and find no open workflows have used `2.2`, but, you actually need `2.2` workers to make progress.

So, seemingly a combination of the version graph and the visibility data is required, which means either fetching both and performing this determination client side (eww, dupe), or creating a new API server side for this purpose (or augmenting the `GetWorkerBuildIDOrdering` api).

We could, of course, postpone this "fully correct" version, since the simpler visibility-only approach is sufficient for most cases

### LINKED ISSUE #4162: [Worker Versioning] Implement API for UI / CLI to know which workers can be retired

### PR DESCRIPTION
**Pull request by bergundy**
- Closes #4162
- Closes #3303

### CODE REVIEW

File: service/matching/matchingEngine.go
Comment (bergundy) [inline service/matching/matchingEngine.go:1224]: Might not be the best place to put this but I couldn't get WFTs dispatched without it.

File: service/frontend/workflow_handler.go
Comment (bergundy) [inline service/frontend/workflow_handler.go:3644]: I think this is desirable and we should avoid leaking error details to the caller.

File: service/frontend/task_reachability.go
Comment (bergundy) [inline service/frontend/task_reachability.go:348]: I think select with limit of one is more efficient, not sure... @rodrigozhou or @alexshtin could verify.

File: service/frontend/task_reachability.go
Comment (bergundy) [inline service/frontend/task_reachability.go:18]: Not sure if this is idiomatic but I found myself repeating this pattern a few times in this file and ended up refactoring to use this little helper. Maybe we have something similar in the codebase or a shared place we can put this.

File: common/persistence/sql/sqlplugin/sqlite/task.go
Comment (bergundy) [inline common/persistence/sql/sqlplugin/sqlite/task.go:378]: Intentional, I started implementing it but realized I can't test it properly and decided to leave this panic as a reminder that this has to be tested.

File: common/persistence/cassandra/matching_task_store.go
Comment (bergundy) [inline common/persistence/cassandra/matching_task_store.go:665]: I'm going to error out here if the number of rows exceeds some (fairly high) dynamic config limit.

File: common/persistence/cassandra/matching_task_store.go
Comment (bergundy) [inline common/persistence/cassandra/matching_task_store.go:663]: We'll need to monitor when this happens and probably prevent another insert.

```suggestion
			if len(taskQueues) == request.Limit {
```

File: service/frontend/task_reachability.go
Comment (dnr) [reply service/frontend/task_reachability.go:223]: very possible the count is more efficient, esp. if it can be approximate. databases often keep summary stats to aid the query planner and the count might be able to be answered from those without touching any data at all

**Review by dnr — COMMENTED** [review]
just a few high-level comments until the new version

File: service/matching/matchingEngine.go
Comment (dnr) [reply service/matching/matchingEngine.go:1228]: The right fix is this: https://github.com/temporalio/temporal/blob/worker-versioning/service/matching/matchingEngine.go#L1208-L1210
If you rebase I think things should work better

File: cmd/tools/rpcwrappers/main.go
Comment (bergundy) [reply cmd/tools/rpcwrappers/main.go:188]: I'll randomize this so it doesn't hit the same node.

File: common/persistence/cassandra/matching_task_store.go
Comment (bergundy) [reply common/persistence/cassandra/matching_task_store.go:663]: Not relevant, validation will happen on write.

File: tests/advanced_visibility_test.go
Comment (Sushisource) [inline tests/advanced_visibility_test.go:1996]: nit: This test reads slightly more obviously if this update happens after the first one so it's both tq1 ops, then tq2, then tq3

**Review by Sushisource — APPROVED** [review]
This LGTM but I'm no DB expert

File: common/dynamicconfig/constants.go
Comment (bergundy) [reply common/dynamicconfig/constants.go:162]: Hmm.. but this is dynamic config that technically can change at any time.
I do get the point though, I don't know if we support dynamic config validation.
I can check.

File: common/persistence/sql/sqlplugin/matching_task_queue.go
Comment (bergundy) [reply common/persistence/sql/sqlplugin/matching_task_queue.go:67]: Hmmm... yeah...
My point was that you're adding to the mapping of build id to task queue.
But if it doesn't read right to you, I can change (in a followup PR because I have a couple stacked on top of this).

File: common/persistence/sql/sqlplugin/sqlite/task.go
Comment (bergundy) [reply common/persistence/sql/sqlplugin/sqlite/task.go:379]: Sure, I didn't think it'd matter much because it's a placeholder for when we do implement deletions.

File: common/util/util.go
Comment (bergundy) [reply common/util/util.go:116]: I don't want to block the goroutines when the receiver exits after the first error.

File: schema/cassandra/temporal/schema.cql
Comment (bergundy) [reply schema/cassandra/temporal/schema.cql:108]: Yes, active is "not deleted", I can clarify.

File: service/frontend/workflow_handler.go
Comment (bergundy) [reply service/frontend/workflow_handler.go:3631]: Sure, couldn't hurt

File: tests/advanced_visibility_test.go
Comment (bergundy) [reply tests/advanced_visibility_test.go:1996]: SGTM.

**Review by bergundy — COMMENTED** [review]
Thanks @Sushisource I'll address your comments after I merge my stacked PRs.

File: common/util/util.go
Comment (Sushisource) [reply common/util/util.go:116]: That's what I mean by using a select on where you take out as well, you can make that nonblocking too.

Comment (bergundy) [reply common/util/util.go:116]: I would need to use select where I send the results.

File: service/frontend/task_reachability.go
Comment (dnr) [inline service/frontend/task_reachability.go:46]: better to give mutexes a name so they're not public.. it's a private concern of this object
```suggestion
	lock sync.Mutex
```

File: service/frontend/task_reachability.go
Comment (dnr) [inline service/frontend/task_reachability.go:49]: comments for wht the keys/values represent?

Comment (dnr) [inline service/frontend/task_reachability.go:49]: why not put the chan in the response struct so you don't need parallel maps?

File: service/frontend/task_reachability.go
Comment (dnr) [inline service/frontend/task_reachability.go:82]: I think I'd find this easier to understand with a single map

File: service/frontend/task_reachability.go
Comment (dnr) [inline service/frontend/task_reachability.go:205]: don't trust %q to use the same types of quote and escaping as our sql parser
```suggestion
		buildIdsFilter = fmt.Sprintf(`BuildIds = "%s"`, common.UnversionedSearchAttribute)
```
(this is a constant so no need to escape, but you could if you want)

File: service/frontend/task_reachability.go
Comment (dnr) [inline service/frontend/task_reachability.go:237]: use quoting and escaping from the sql parser, not from go

File: service/frontend/task_reachability.go
Comment (dnr) [inline service/frontend/task_reachability.go:220]: kind of annoying to duplicate this from the matching code.. can you at least put it in a separate function in this file and we can combine it later?

File: service/frontend/task_reachability.go
Comment (dnr) [inline service/frontend/task_reachability.go:239]: use `searchattribute.BuildIds`

File: service/frontend/task_reachability.go
Comment (dnr) [inline service/frontend/task_reachability.go:271]:
```suggestion
		statusFilter = ` AND ExecutionStatus = "Running"`
```

File: service/frontend/task_reachability.go
Comment (dnr) [inline service/frontend/task_reachability.go:288]: use constant for search attribute name

File: service/matching/matchingEngine.go
Comment (dnr) [inline service/matching/matchingEngine.go:126]: consider using `common/locks.IDMutex` here. it takes some of the boilerplate out of this pattern. you can use separate keys for the update + replication locks

File: proto/internal/temporal/server/api/matchingservice/v1/service.proto
Comment (dnr) [inline proto/internal/temporal/server/api/matchingservice/v1/service.proto:99]: this sounds too tempting to use when you want to update user data. can we make the comment very explicit that these two functions should only be called by the matching implementation itself and that external callers like frontend should use `UpdateWorkerBuildIdCompatibility` (or some other function for other fields)?

it might even be good if the name starts with `Internal`

File: service/matching/taskQueueManager.go
Comment (dnr) [inline service/matching/taskQueueManager.go:491]: remove this now that lint won't complain?

File: cmd/tools/rpcwrappers/main.go
Comment (bergundy) [reply cmd/tools/rpcwrappers/main.go:189]: Right

File: common/persistence/sql/sqlplugin/matching_task_queue.go
Comment (bergundy) [reply common/persistence/sql/sqlplugin/matching_task_queue.go:67]: like

File: common/persistence/sql/sqlplugin/matching_task_queue.go
Comment (bergundy) [reply common/persistence/sql/sqlplugin/matching_task_queue.go:73]: like

File: common/persistence/sql/sqlplugin/matching_task_queue.go
Comment (bergundy) [reply common/persistence/sql/sqlplugin/matching_task_queue.go:84]: I will, just curious why

File: common/persistence/sql/sqlplugin/mysql/task.go
Comment (bergundy) [reply common/persistence/sql/sqlplugin/mysql/task.go:361]: It should never happen, we don't support deletion, and when we do there's no way we won't catch this.
I almost regret adding this to the codebase.

File: common/persistence/sql/sqlplugin/sqlite/task.go
Comment (bergundy) [reply common/persistence/sql/sqlplugin/sqlite/task.go:369]: Yeah, can do, just don't have a way to test this.

File: common/util/util.go
Comment (bergundy) [reply common/util/util.go:115]: Yes but I find it more readable to have a multiline inline function as the last param.
Go is probably the least functional modern language.

File: common/worker_versioning.go
Comment (bergundy) [reply common/worker_versioning.go:25]: I'll leave this up to you to decide, I don't have a strong opinion, you can move this and your stuff there too.
Seemed like common is where anything that needs to be shared in the server goes 😬.

File: common/worker_versioning.go
Comment (bergundy) [reply common/worker_versioning.go:31]: fancy

File: common/worker_versioning.go
Comment (bergundy) [reply common/worker_versioning.go:47]: I think this my name is better, it's just a conversion function between these 2 types. And there's no overloading in Go so if we ever would want to convert something else to the search attribute the `Make` function name would already be taken.

File: proto/internal/temporal/server/api/matchingservice/v1/service.proto
Comment (bergundy) [reply proto/internal/temporal/server/api/matchingservice/v1/service.proto:99]: Can change to `Internal`

File: schema/cassandra/temporal/schema.cql
Comment (bergundy) [reply schema/cassandra/temporal/schema.cql:105]: I like the explicit and long name TBH.

File: schema/cassandra/temporal/schema.cql
Comment (bergundy) [reply schema/cassandra/temporal/schema.cql:108]: Good catch!

File: service/frontend/task_reachability.go
Comment (bergundy) [reply service/frontend/task_reachability.go:46]: ok

File: service/frontend/task_reachability.go
Comment (bergundy) [reply service/frontend/task_reachability.go:49]: I'll change this to use futures

File: service/frontend/task_reachability.go
Comment (bergundy) [reply service/frontend/task_reachability.go:220]: Put this in your favorite place (common), we can move it out later.

File: service/frontend/task_reachability.go
Comment (bergundy) [reply service/frontend/task_reachability.go:237]: I thought the Go one was safe, I guess not...
Do you have an example I for reference?

File: common/persistence/sql/sqlplugin/matching_task_queue.go
Comment (dnr) [reply common/persistence/sql/sqlplugin/matching_task_queue.go:84]: otherwise you could pass a value of one type as the other? I mean, yeah, it's not like it's a big problem, they are the same in some sense, it's just weird to be able to pass a `*CountTaskQueuesByBuildIdRequest` to `GetTaskQueuesByBuildId`

File: common/persistence/sql/sqlplugin/mysql/task.go
Comment (dnr) [reply common/persistence/sql/sqlplugin/mysql/task.go:361]: Someone implements this for sqlite, writes some code to call it, tests locally, doesn't write an integration test, it somehow gets released, it gets called in production, boom. We shouldn't be in the habit of putting panics in places where we could just as easily return NotImplemented

File: common/persistence/sql/sqlplugin/sqlite/task.go
Comment (dnr) [reply common/persistence/sql/sqlplugin/sqlite/task.go:369]: can leave a TODO and do it later, obviously it's not useful yet

File: common/util/util.go
Comment (dnr) [reply common/util/util.go:115]: fair

File: schema/cassandra/temporal/schema.cql
Comment (dnr) [reply schema/cassandra/temporal/schema.cql:105]: what if you want to define another "index" by a different field, so now the name is wrong because there are two different types of index?

also this is nitpicky but, the "row" is not "an index", it's an entry in an index. so the name isn't even really correct

File: service/frontend/task_reachability.go
Comment (dnr) [reply service/frontend/task_reachability.go:237]: we use `github.com/xwb1989/sqlparser` for parsing queries. disappointingly there's no simple "quote" function but I figured out you can do it like this: https://go.dev/play/p/MzUip_xtkiM

File: service/matching/matchingEngine.go
Comment (bergundy) [reply service/matching/matchingEngine.go:126]: I think this is fine TBH

**Pull request merged**
